lib/semantics/check-do.cc

```diff
@@ -395,6 +395,20 @@ class DoContext {
     }
   }
 
+  // Step expressions can't be zero (Section 11.1.7.4.1, paragraph 1).
+  // Make this a warning since it's not a constraint violation
+  void CheckZeroStep(const parser::ScalarExpr &stepExpression) {
+    MaybeExpr stepExpr{evaluate::Fold(
+        context_.foldingContext(), AnalyzeExpr(context_, stepExpression))};
+    if (stepExpr) {
+      std::optional<std::int64_t> stepValue{evaluate::ToInt64(*stepExpr)};
+      if (stepValue && stepValue.value() == 0) {
+        context_.Say(stepExpression.thing.value().source,
+            "DO step expression should not be zero"_en_US);
```

Comment:
Constraint violations and program requirement violations are both equally bad (Clause 4.2).  The difference is that the processor (compiler) is *required* to diagnose the former, but not the latter.  This is usually because it is difficult or impossible to diagnose all program requirement violations, not because they are necessarily less problematic.  In cases where as an aid to the programmer it *is* possible to diagnose a program requirement violation, the violation should usually be an error, just like a constraint.  In the absence of an -Mstandard option, both constraint and program requirement violations could possibly appear as warnings rather than errors if the violation doesn't necessarily create a problem with how the program should be interpreted.  In this case, if a step value of zero is likely to result in an infinite loop, that seems like it should be a hard error.

Reply:
Thanks for the clarification, Val.  I'll update the comment.

Reply:
You might have
```
integer, parameter :: debug_step = 0
if(debug_step > 0) then
  do i = 1, N, debug_step  !  Want compilation not to fail
```

Reply:
I've searched the standard, and I found 12 places where prohibit zero values for various constructs -- 
Section 9.5.3.3.2, paragraph 2
Section 11.1.7.4.1, paragraph 1, point (1)
Section 11.1.7.4.2, paragraph 1, point (1)
Section 13.7.2.1, paragraph 6
Section 13.7.2.2, paragraph 1
Section 13.7.2.3.2, paragraph 1
Section 13.7.2.3.6, paragraph 2
Section 13.7.2.4, paragraph 1
Section 13.7.5.1, paragraph 1
Section 16.9.118, paragraph 3
Section 16.9.135, paragraph 3
Section 16.9.136, paragraph 3
I agree that the step expression for a normal DO construct is a context where a warning is preferable to an error.  What about these other contexts, though?  For example, what about the step expression for a DO CONCURRENT?

Reply:
A program violation, if diagnosed, should generally be an error.

Reply:
Thanks for the feedback, Val.  I have yet to look at these in detail, but my default will be to call them errors, starting with DO CONCURRENT.